collectors/hooks.php

```diff
@@ -14,6 +14,58 @@ public function name() {
 		return __( 'Hooks & Actions', 'query-monitor' );
 	}
 
+	public function __construct() {
+		parent::__construct();
+		add_action( 'qm/listen/start', array( $this, 'action_function_listen_start' ), 10, 1 );
+		add_action( 'qm/listen/stop',  array( $this, 'action_function_listen_stop' ), 10, 1 );
+	}
+
+	public function action_function_listen_start( $label ) {
+		if ( ! array_key_exists( 'discovered_hooks', $this->data ) ) {
+			$this->data['discovered_hooks'] = array();
+		}
+
+		if ( array_key_exists( $label, $this->data['discovered_hooks'] ) ) {
+			return;
+		}
+
+		$this->data['discovered_hooks'][ $label ] = array();
+
+		add_action( 'all', array( $this, 'action_function_listen_all' ), 0, 1 );
+	}
+
+	public function action_function_listen_all( $var = null ) {
+		if ( in_array( current_action(), array(
+			'qm/listen/start',
+			'qm/listen/stop',
+		) ) ) {
+			return $var;
+		}
+
+		global $wp_actions;
+
+		end( $this->data['discovered_hooks'] );
+		$label = key( $this->data['discovered_hooks'] );
+		$last = end( $this->data['discovered_hooks'][ $label ] );
+
+		if ( current_action() === $last['hook'] ) {
+			$i = key( $this->data['discovered_hooks'][ $label ] );
+			$this->data['discovered_hooks'][ $label ][ $i ]['count']++;
+		} else {
+			$this->data['discovered_hooks'][ $label ][] = array(
+				'is_action' => array_key_exists( current_action(), $wp_actions ),
```

Comment:
Ohhhhhh, nice. I've spent a bit of time looking into global hook timing and hadn't found a way to distinguish between actions and filters without fetching a stack trace, which adds too much overhead. I really want to re-visit timings for individual callbacks on actions, and this might just allow me to do that in a performant way. Thanks.